### PR TITLE
Distinguish between /regexp/ and "regexp" in stringified output

### PR DESCRIPTION
**Pull request by ypdn**
Fixes issue #221

### CODE REVIEW

File: internal/ast/ast.go [hunk 1]
```diff
@@ -292,12 +292,17 @@ func (e *NumExpr) String() string {
...
 type StrExpr struct {
-	Value string
+	Value  string
+	Regexp bool
 }
 
 func (e *StrExpr) String() string {
+	if e.Regexp {
+		r := &RegExpr{e.Value}
+		return r.String()
```
Comment (benhoyt) [inline internal/ast/ast.go:304]: Nit suggestion (untested, but I think it should work):
```suggestion
		return RegExpr.String(e.Value)
```

File: internal/ast/ast.go [hunk 2]
```diff
@@ -292,12 +292,17 @@ func (e *NumExpr) String() string {
 	}
 }
 
-// StrExpr is a literal string like "foo".
+// StrExpr is a literal string like "foo" or a regexp constant like /foo/.
 type StrExpr struct {
-	Value string
+	Value  string
+	Regexp bool
```
Comment (benhoyt) [inline internal/ast/ast.go:298]: Naming nit: can we call this `IsRegexp` for clarity?

**Review by benhoyt — CHANGES_REQUESTED** [review]
Nice -- this is a very simple and tidy fix! I hadn't thought of that approach.

Just a couple of nit suggestions, and can we please add a test? You should be able to just add a line or two to the existing `TestParseAndString` test in `parser_test.go`.

File: internal/ast/ast.go (hunk 1 above)
Comment (ypdn) [reply internal/ast/ast.go:304]: It won't, because the (*RegExpr).String takes a *RegExpr as an argument, not a string.

Comment (benhoyt) [reply internal/ast/ast.go:304]: Ah yes, you're right.

Comment (benhoyt) [reply internal/ast/ast.go:304]: Building a `RegExpr` object here still feels a little weird -- can we factor out the content of `RegExpr.String` into `formatRegex` and call it from both places?

Comment (ypdn) [reply internal/ast/ast.go:304]: Yeah

File: internal/ast/ast.go (hunk 2 above)
Comment (ypdn) [reply internal/ast/ast.go:298]: We could, but that's so not-Go

Comment (ypdn) [reply internal/ast/ast.go:298]: Maybe we should document it better?

Comment (benhoyt) [reply internal/ast/ast.go:298]: Looks like your mostly right that it's "not Go", though not entirely, [for example, the `Is*` fields here](https://pkg.go.dev/net#DNSError). However, we have another bool in this package, `IncrExpr.Pre`, that doesn't have the `Is`, so let's go with that.

However, on closer look, can we please call it `Regex bool`? The "regexp" with a 'p' kind of refers to Go's "regexp" package, whereas we I tend to call it "regex" in an AWK context (for example, the `RegExpr.Regex` field itself).

Comment (ypdn) [reply internal/ast/ast.go]: Done

**Review by benhoyt — APPROVED** [review]
Nice, thanks for the contribution!

**Comment by ypdn** [comment]
@benhoyt Could you create a new tag so that `go get` would download it?

**Comment by benhoyt** [comment]
That's okay: 1) the `ast` package is in `internal` so it can't be used outside of this repo, and 2) Go package compatibility usually doesn't guarantee that will continue working -- see "struct literals" in https://go.dev/doc/go1compat

**Pull request merged**